client/src/templates/Challenges/classic/mobile-layout.tsx

```diff
@@ -45,6 +45,69 @@ class MobileLayout extends Component<MobileLayoutProps, MobileLayoutState> {
     });
   };
 
+  avoidWindowScrolling = (event: Event) => {
+    event.preventDefault();
+  };
+
+  setToolPanelPosition = () => {
+    const toolPanelGroup = (
+      document.getElementsByClassName(
+        'tool-panel-group-mobile'
+      ) as HTMLCollectionOf<HTMLElement>
+    )[0];
+
+    if (navigator.userAgent.match(/iPhone|Android.+Mobile/)) {
+      if (
+        visualViewport?.height &&
+        window.innerHeight > visualViewport.height
+      ) {
+        setTimeout(() => {
+          window.scrollTo(0, 0);
+
+          toolPanelGroup.style.position = 'absolute';
+          if (visualViewport?.height !== undefined) {
+            toolPanelGroup.style.bottom =
+              String(window.innerHeight - visualViewport.height) + 'px';
+          }
+        }, 200);
+
+        window.addEventListener('touchmove', this.avoidWindowScrolling, {
+          passive: false
+        });
+      } else {
+        document.documentElement.style.height =
+          String(window.innerHeight) + 'px';
+
+        toolPanelGroup.style.position = '';
+        toolPanelGroup.style.bottom = '';
+
+        window.removeEventListener('touchmove', this.avoidWindowScrolling);
+      }
+    }
+  };
+
+  componentDidMount() {
+    if (navigator.userAgent.match(/iPhone|Android.+Mobile/)) {
+      window.addEventListener('resize', this.setToolPanelPosition);
+      visualViewport?.addEventListener('resize', this.setToolPanelPosition);
```

Comment:
Do we need to add event listeners to the window and the visaulViewport?

Reply:
If you are adding the window event listener to fix the issue with the address bar only, you could separate that into its own function.

Reply:
Sorry, to anchor the tool panel buttons at the bottom, it seems we can achieve this with only visualViewport resize event listener because the event listener could be used both to detect opening address bar and appearing of the virtual keyboard. So window event listener is redundant.